hazelcast/src/main/java/com/hazelcast/query/impl/predicates/EvaluateVisitor.java

```diff
@@ -0,0 +1,266 @@
+            }
+            List<EvaluatePredicate> group = evaluable.get(indexName);
+            if (group == null) {
+                group = new ArrayList<EvaluatePredicate>(predicates.length);
+                evaluable.put(indexName, group);
+            } else {
+                requiresGeneration = true;
+            }
+            group.add(evaluatePredicate);
+        }
+
+        if (!requiresGeneration) {
+            // no changes to the predicates required
+            return orPredicate;
+        }
+
+        // Add non-evaluable predicates to the output.
+
+        List<Predicate> output = new ArrayList<Predicate>();
+        for (Predicate subPredicate : predicates) {
+            if (!(subPredicate instanceof EvaluatePredicate)) {
+                output.add(subPredicate);
+                continue;
+            }
+
+            EvaluatePredicate evaluatePredicate = (EvaluatePredicate) subPredicate;
+            String indexName = evaluatePredicate.getIndexName();
+            Index index = indexes.matchIndex(indexName, QueryContext.IndexMatchHint.EXACT_NAME);
+            if (!index.canEvaluate(orPredicate.getClass())) {
+                output.add(subPredicate);
+            }
+        }
+
+        // Add evaluable predicates to the output.
+
+        for (Map.Entry<String, List<EvaluatePredicate>> groupEntry : evaluable.entrySet()) {
+            String indexName = groupEntry.getKey();
+            List<EvaluatePredicate> group = groupEntry.getValue();
+
+            if (group.size() == 1) {
+                output.add(group.get(0));
+                continue;
+            }
+
+            Predicate[] groupPredicates = new Predicate[group.size()];
+            for (int i = 0; i < groupPredicates.length; ++i) {
+                groupPredicates[i] = group.get(i).getPredicate();
+            }
+            output.add(new EvaluatePredicate(new OrPredicate(groupPredicates), indexName));
+        }
+
+        return output.size() == 1 ? output.get(0) : new OrPredicate(output.toArray(EMPTY_PREDICATES));
+    }
+
+    @Override
+    public Predicate visit(NotPredicate notPredicate, Indexes indexes) {
+        Predicate subPredicate = notPredicate.getPredicate();
+        if (!(subPredicate instanceof EvaluatePredicate)) {
+            return notPredicate;
+        }
+
+        EvaluatePredicate evaluatePredicate = (EvaluatePredicate) subPredicate;
+        String indexName = evaluatePredicate.getIndexName();
+        Index index = indexes.matchIndex(indexName, QueryContext.IndexMatchHint.EXACT_NAME);
+        if (!index.canEvaluate(notPredicate.getClass())) {
+            return notPredicate;
+        }
+
+        return new EvaluatePredicate(new NotPredicate(evaluatePredicate.getPredicate()), indexName);
+    }
+
+    @Override
+    public Predicate visit(EqualPredicate predicate, Indexes indexes) {
+        Index index = indexes.matchIndex(predicate.attributeName, QueryContext.IndexMatchHint.PREFER_UNORDERED);
```

Comment:
Probably we should always prefer bitmap indexes over hash ones. If user adds a bitmap index he or she is doing that for a reason. The only question what is the purpose of hash index existence in that case, the index would never be used for actual queries.

It makes sense to prefer hash index over bitmap index for certain queries (like plain `a = 10`), since it's faster to do a hash table lookup rather than iterating a bitmap. But with the current query optimizer implementation it's hard to introduce such a preference. Predicate visitor should know that `a = 10` is not embedded in some non-terminal predicate (`and`, `or`, `not`), but visitors may know only the current predicate and its children. And cardinality also matters :)

Reply:
Makes sense, thank you for explanation.

Reply:
fixed in https://github.com/hazelcast/hazelcast/pull/16162/commits/940eb038245403f3715fa872810bc3bf29f1b553